### PR TITLE
MINOR: Avoid unnecessary allocations in index binary search

### PR DESCRIPTION
**Pull request by ijuma**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java [hunk 1]
```diff
@@ -484,27 +478,35 @@ private static MappedByteBuffer createMappedBuffer(RandomAccessFile raf, boolean
...
         // check if the target offset is smaller than the least offset
-        if (compareIndexEntry(parseEntry(idx, 0), target, searchEntity) > 0)
-            return new BinarySearchResult(-1, 0);
+        if (compareIndexEntry(parseEntry(idx, 0), target, searchEntity) > 0) {
+            switch (searchType) {
+                case SMALLEST_UPPER_BOUND:
+                    return -1;
+                case LARGEST_LOWER_BOUND:
+                    return 0;
+            }
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java:503]: How about throwing `IllegalStateException` from the default label for consistency?

File: storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java [hunk 2]
```diff
@@ -484,27 +478,35 @@ private static MappedByteBuffer createMappedBuffer(RandomAccessFile raf, boolean
...
+            return binarySearch(idx, target, searchEntity, searchType, firstHotEntry, entries - 1);
         }
 
         // check if the target offset is smaller than the least offset
-        if (compareIndexEntry(parseEntry(idx, 0), target, searchEntity) > 0)
-            return new BinarySearchResult(-1, 0);
+        if (compareIndexEntry(parseEntry(idx, 0), target, searchEntity) > 0) {
+            switch (searchType) {
+                case SMALLEST_UPPER_BOUND:
+                    return -1;
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java]: It seems to me `SMALLEST_UPPER_BOUND` should return 0 and `LARGEST_LOWER_BOUND` should return -1

File: storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java [hunk 3]
```diff
@@ -517,13 +519,19 @@ private BinarySearchResult binarySearch(ByteBuffer idx, long target, IndexSearch
             else if (compareResult < 0)
                 lo = mid;
             else
-                return new BinarySearchResult(mid, mid);
+                return mid;
+        }
+        switch (searchType) {
+            case SMALLEST_UPPER_BOUND:
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java]: ditto. maybe we get the incorrect returned value.

**Review by chia7712 — COMMENTED** [review]
@ijuma nice patch. A couple of minor comments below. PTAL

**Review by divijvaidya — APPROVED** [review]

**Comment by divijvaidya** [comment]
^ please ignore my approval. @chia7712's comments need to be fixed first.

File: storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java (hunk 1 above)
Comment (ijuma) [reply storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java:503]: It's better to have exhaustive matches than to add a `default`. Unfortunately, the Java compiler doesn't always infer that the method always returns due to an exhaustive match (at least not with old school switches, the new switch probably handles it better) and hence we have to add the explicit `default`. So, I'd prefer to leave as is.

File: storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java (hunk 2 above)
Comment (ijuma) [reply storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java]: Good catch, I reversed it here and in the other place. Will fix both.

**Comment by ijuma** [comment]
@chia7712 Fixed the bug and also tweaked the name of the enum and variable to avoid confusion with `SearchEntityType`.

**Comment by chia7712** [comment]
> Fixed the bug and also tweaked the name of the enum and variable to avoid confusion with SearchEntityType.

Are there failed tests related to that bug (inverse usage of SearchType) ? Not sure why I don't see the QA triggered by this PR

**Comment by ijuma** [comment]
> Are there failed tests related to that bug (inverse usage of SearchType) ? Not sure why I don't see the QA triggered by this PR

I was wondering the same thing for both and am investigating. :)

**Comment by ijuma** [comment]
@chia7712 `OffsetIndexTest` does fail without the fix and pass with the fix. And the CI build is queued now.

**Comment by ijuma** [comment]
Test failures are unrelated:

> Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testReplicationIsCreatingTopicsUsingProvidedForwardingAdmin()
> Build / JDK 11 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testDelayedConfigurationOperations()
> Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationSSLTest.testReplication()
> Build / JDK 8 and Scala 2.12 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testCreatePartitionsUseProvidedForwardingAdmin()
> Build / JDK 8 and Scala 2.12 / kafka.api.SslEndToEndAuthorizationTest.[2] quorum=kraft, isIdempotenceEnabled=false
> Build / JDK 8 and Scala 2.12 / kafka.api.TransactionsExpirationTest.testTransactionAfterProducerIdExpires(String).quorum=kraft
> Build / JDK 8 and Scala 2.12 / kafka.server.FetchRequestBetweenDifferentIbpTest.testControllerOldIBP()
> Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.SmokeTestDriverIntegrationTest.[1] true

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by ijuma** [comment]
Annoyingly, the commit was somehow pushed prematurely while I was still editing the commit message - not quite sure how that happened. Apologies for the bad commit message and the missing reviewers line.

It would be quite handy to have an `undo` for a few seconds like Gmail's.

**Pull request merged**